README.md
# BBVisa2OFX

[![Build Status](https://travis-ci.org/josecoelho/bbvisa2ofx.svg?branch=master)](https://travis-ci.org/josecoelho/bbvisa2ofx)

This software converts the file of credit card accounts, generated from
Banco do Brasil website, to the OFX format.

You are not from Brazil? So you do not has an account on Banco do Brasil
and this software is useless for you... But if you want a help to
convert a txt file to OFX format (in python), this project would be used
as reference :)

## Portugues-BR:

Este software converte o extrato de cartão de crédito, gerado através do
site do Banco do Brasil, para o formato OFX.

Bem... Eu tenho uma conta no Banco do Brasil e utilizo o GnuCash? para
cuidar das minhas finanças.

Eu sempre exporto os meus extratos para o formato OFX e importo no
GnuCash?. Porém, para o extrato do cartão de credito, VISA no meu caso,
o Banco do Brasil só oferece a opção de exportar para TXT. Então, para
evitar fadiga, eu criei este "script" em python que converte este
arquivo TXT para OFX.

Agora posso importar o extrato do Cartão de credito do Banco do Brasil
no GnuCash? e não perder tempo digitando linha a linha.

Cartoes Master Card tambem sao suportados, e voce pode importar varios
cartoes diferentes que o gnucash vai considerar contas distintas.
(agradecimentos ao Daniel Roviriego)

Mais uma contribuição. Graças ao patch disponibilizado por Rodrigo
Primo, agora o valores em dolar são automaticamente convertidos para
real utilizando a taxa de conversão disponível na fatura.

## INSTALL


```
pip install bbvisa2ofx
```

## Running


With GUI
```
	bbvisa2ofx
```

Commandline
```
    bbvisa2ofx_cli
```


## Development

### Contribution

1. Fork it
2. Create your feature branch (git checkout -b my-new-feature)
3. Commit your changes (git commit -am 'Add some feature')
4. Push to the branch (git push origin my-new-feature)
5. Create new Pull Request


### Running tests

From root dir exec
```
 python -m bbvisa2ofx.test.txtparser
```

### Publishing a new release (I think it can be done only by me...)

1. increment the version number in your setup.py file,
2. update your CHANGES.txt file,
3. run python setup.py sdist upload again.


WEB Interface
==============

Web Interface on https://github.com/josecoelho/bbvisa2ofx_web




CHANGES.txt
v0.21, 2014-11-26 -- Installment transactions dates are moved X months ahead (X = Installment number - 1)
v0.20, 2014-11-25 -- Guess year from transactions using due date as reference
v0.18, 2014-09-01 -- Change date format of txt from dd/mm/yy to dd/mm
v0.18, 2014-08-15 -- Noobie packaging try :p
v0.17b, 2014-08-15 -- Packaging to run with pip install


bbvisa2ofx/test/txtparser.py
'''
Created on Jun 9, 2010

@author: coelho
'''
import unittest
from bbvisa2ofx.txtparser import TxtParser
from bbvisa2ofx.bbvisa2ofx import convert
import os, datetime

class Test(unittest.TestCase):
    file_path = os.path.dirname(os.path.abspath(__file__))+"/exemploFaturaCartao.txt";
    parser = ''

    def setUp(self):
        fTxt = open(self.file_path,'r')

        self.parser = TxtParser(fTxt);

    def testParseTransactionLine(self):
        self.parser.dueDate = datetime.datetime(2011,12,25)
        parsedLine = self.parser.parseTransactionLine('11/08    PGTO DEBITO CONTA 3333 000006037  200  211        -1.000,00        0,00')
        self.assertEquals(parsedLine['date'],'20110811')
        self.assertEquals(parsedLine['fitid'],'201108111000.0PGTODEBITOCONTA3333000006037200')
        self.assertEquals(parsedLine['value'],1000.00)
        self.assertEquals(parsedLine['desc'],'PGTO DEBITO CONTA 3333 000006037  200  ')

    def testParseValueFromTransactionLine(self):
        line = '30/07    NETFLIX.COM            SAO PAULO       BR             16,90        0,00'
        value = self.parser.parseValueFromTransactionLine(line)

        self.assertEquals(value, -16.90)

    def testParseValueFromTransactionLine_ConvertingDollarToReal(self):
        line = '25/07    Atlassian              Sydney          AU              0,00       20,00'
        self.parser.exchangeRate = 2 #force exchangeRate
        value = self.parser.parseValueFromTransactionLine(line)

        self.assertEquals(value, -40.00)

    def testParseDateFromTransactionLine_fromSameYear(self):
        line = '25/02    Atlassian              Sydney          AU              0,00       20,00'
        self.parser.dueDate = datetime.datetime(2011,07,25)

        date = self.parser.parseDateFromTransactionLine(line)
        self.assertEquals(date,'20110225')

    def testParseDateFromTransactionLine_guessingYearBefore(self):
        '''
            mesmo mes do vencimento, sugerindo ano anterior
        '''
        line = '25/07    Atlassian              Sydney          AU              0,00       20,00'
        self.parser.dueDate = datetime.datetime(2011,07,25)

        date = self.parser.parseDateFromTransactionLine(line)
        self.assertEquals(date,'20100725')

    def testParse(self):
        self.parser.parse()
        # print self.parser.items
        self.assertEquals(14, len(self.parser.items) )

    def testParse_ConvertingDollarToReal(self):
        self.parser.parse()
        self.assertEquals(-46.86, self.parser.items[1]['value'])

    def testConvert(self):
        fTxt = open(self.file_path,'r')
        fOfx = open(self.file_path+".ofx",'w')

        convert(fTxt,fOfx)

    def testIsDolarLine(self):
        self.assertEquals(self.parser.parseExchangeRateLine('      0,00 -       0,00 +       0,00 =        0,00   X      0.0 =           0,00'),0.0)
        self.assertEquals(self.parser.parseExchangeRateLine('    871,64 -      21,78 +      69,52 =      823,90   X   1.7627 =       1.452,28'),1.7627)
        self.assertEquals(self.parser.parseExchangeRateLine('  1.172,33 -   1.172,33 +   1.039,98 =    1.039,98 -       0,00 =       1.039,98'),0.0)
        self.assertEquals(self.parser.parseExchangeRateLine('    214,17 -     214,79 +     162,00 =      161,38 -       0,00 =         161,3'),0.0)
        self.assertEquals(self.parser.parseExchangeRateLine('11/08/10 TARIFA SOBRE COMPRAS NO EXTERIOR                       0,00        0,44'),0.0)
        self.assertEquals(self.parser.parseExchangeRateLine(''),0.0)


    def testUpdateInstallmentTransactionLine(self):
        obj = {}
        obj['value'] = 20.0
        obj['date'] = '20110504'
        obj['desc'] = 'ASDASDASDASDA PARC 02/04 FARROUPILHA   BR'
        obj['fitid'] = (obj['date'] + str(obj['value']) + obj['desc']).replace(' ','')

        self.parser.dueDate = datetime.datetime(2011,07,25)
        self.parser.updateDateFromInstallmentTransactionLine(obj)
        self.assertEquals(obj['date'], '20110604')

    def testParseInstallmentTransactionLine(self):
        self.parser.dueDate = datetime.datetime(2011,07,25)
        firstInstallmentTx = self.parser.parseTransactionLine('04/05    ASDASDASDASDA PARC 01/04 FARROUPILHA   BR            141,58        0,00')
        self.assertEquals(firstInstallmentTx['date'], '20110504')
        secondInstallmentTx = self.parser.parseTransactionLine('04/05    ASDASDASDASDA PARC 02/04 FARROUPILHA   BR            141,58        0,00')
        self.assertEquals(secondInstallmentTx['date'], '20110604')

if __name__ == "__main__":
    #import sys;sys.argv = ['', 'Test.testName']
    unittest.main()


bbvisa2ofx/txtparser.py
# -*- coding: cp1252 -*-
'''
Created on Jun 9, 2010

@author: coelho
'''

import re
from datetime import datetime
from dateutil.relativedelta import relativedelta

class TxtParser:
    '''
    Classe responsavel por realizar o parse do arquivo txt da fatura de cartoes
    disponibilizada pelo banco do brasil.

    Analizando o arquivo disponibilizado, verifica-se que as linhas que interessam sao as que possuem uma
    data no formato dd/mm no inicio.

    Para cada linha contendo este padrao, vamos extrair as informacoes da seguinte forma:
        date: primeiros 5 caracteres
        desc: do caracter 8 ao 49
        value:
            split no final da linha do 51 caracter em diante, primeiro item
    '''
    items = None
    cardTitle = None #titulo do cartao, definido por "Modalidade" no txt
    cardNumber = None #numero do cartao, definido por Nr.Cartao no txt
    txtFile = None
    dueDate = None #data de vencimento, definido por "Vencimento" no txt

    def __init__(self,txtFile):
        '''
        Constructor
        '''
        self.items = []
        self.txtFile = txtFile
        self.exchangeRate = 0.0

    def parse(self):
        f = self.txtFile
        lines = f.readlines()

        for line in lines:
            self.parseDueDate(line)
            self.parseExchangeRateLine(line)
            self.parseCardTitleLine(line)
            self.parseCardNumberLine(line)

        #now with the exangeRate and dueDate populated, we can parse all transaction lines
        for line in lines:
            self.parseTransactionLine(line)

    def parseDueDate(self, line):
        '''
        popula dueDate se for a linha que representa o vencimento da fatura,
        esta informacao eh utilizada para adivinharmos o ano da compra
        FIXME: ainda pode gerar problemas quando temos uma compra realizada no ano anterior, mas que aparenta ser do ano atual

        '''

        if(line.lstrip().startswith("Vencimento")):
                print "Due date line found. %s" % line

                self.dueDate = datetime.strptime(line.split(":")[1].strip(),'%d.%m.%Y')
                print "Due date is: %s" % self.dueDate


    def parseExchangeRateLine(self, line):
        '''
        popula exchangeRate se for a linha que apresenta o resumo dos gastos
        em dolar. essa linha eh utilizada para extrair o valor da taxa
        de conversao de dolar para real. o que diferencia essa linha da
        linha com o resumo dos gastos em reais a presenca de um sinal
        de multiplicacao (representado por um X)
        '''

        if (re.match('^\s+\S+\s+-\s+\S+\s+\+\s+\S+\s+=\s+\S+\s+X', line)):
            print "Echange Rate line found: "+ line
            rate = re.findall('X\s+(\S+)', line)[0]
            rate = rate.replace(',','.')
            self.exchangeRate = float(rate)
            print "Exchange Rate value: "+str(self.exchangeRate)
            return self.exchangeRate

        return 0.0


    def parseCardTitleLine(self, line):
        '''
            Titulo do cartao inicia com "Modalidade"
        '''

        if(line.lstrip().startswith("Modalidade")):
                print "Card title line found. %s" % line

                # Fix issue #3 changing from lstrip to strip
                self.cardTitle = line.split(":")[1].strip()
                print "The card title is: %s" % self.cardTitle

    def parseCardNumberLine(self, line):
        '''
            Numero do cartao inicia com "Nr.Cart"
        '''

        if(line.lstrip().startswith("Nr.Cart")):
                print "Card number line found. %s" % line
                # Fix issue #3 changing from lstrip to strip
                self.cardNumber = line.split(":")[1].strip()
                print "The card number is: %s" % self.cardNumber



    def parseTransactionLine(self, line):
        '''
        Linhas de transacao inicial com uma data no formato "dd/mm "
        (devemos verificar o espaco no final pois existem linhas no formato dd/mm/yyyy que nao sao tansacoes)

        caso for uma linha de transacao, um objeto sera adicionado na lista self.items
        este objeto contem os seguintes campos:
            date: data da transacao
            desc: descricao
            value: valor em BRL
        '''

        if(re.match("^\d\d\/\d\d\ $", line[:6]) != None):
            brlValue = ''
            usdValue = ''

            obj = {}
            obj['value'] = self.parseValueFromTransactionLine(line)
            obj['date'] = self.parseDateFromTransactionLine(line)
            obj['desc'] = line[9:48].lstrip().replace('*','')
            obj['fitid'] = (obj['date'] + str(obj['value']) + obj['desc']).replace(' ','')
            print "Line parsed: "+ str(obj)

            # Atualiza data das transacoes de compras parceladas
            self.updateDateFromInstallmentTransactionLine(obj)

            self.items.append(obj)
            return obj

    def parseValueFromTransactionLine(self, line):
        '''
        Extraindo valor da linha, a partir do caracter 51
        Caso esteja em dolar, converter para real utilizando a taxa de cambio.

        Agradecimento especial para Rodrigo que contribuiu pelo code.google
        '''
        value = 0.0

        arr = line[51:].split()

        brlValue = float(arr[0].replace('.','').replace(',','.'))
        usdValue = float(arr[1].replace('.','').replace(',','.'))

        if brlValue != 0.0:
            value = brlValue
        else:
            value = usdValue * self.exchangeRate

        value = value * -1 #inverte valor

        return value

    def parseDateFromTransactionLine(self, line):
        '''
        Extraindo data da linha de transacao
        Como o BB removeu o ano das datas, vamos precisar "adivinhar" o ano correto,
        conforme a data de vencimento

        definimos o ano do vencimento como padrao, porem caso a data da transacao
        fique maior que o vencimento, assumimos o ano anterior como ano correto.

        FIXME transacoes feitas a mais de 12 meses terao o ano definido incorretamente, mas
        nao consegui pensar em outra solucao no momento

        Agradecimento especial a Leonardo F. Cardoso que fez esta contruicao por email. :)
        '''

        transactionDate = datetime.strptime(line[:5],'%d/%m')
        transactionDate = transactionDate.replace(self.dueDate.year)
        if transactionDate >= self.dueDate:
            transactionDate = transactionDate.replace(transactionDate.year-1)

        return transactionDate.strftime('%Y%m%d')

    def updateDateFromInstallmentTransactionLine(self, obj):
        '''
        Verifica se trata-se de uma linha de transacao de compra parcelada (ex.: PARC 01/04) e,
        em caso positivo, posterga o data de vencimento X meses para frente (X = Nro Parc - 1)
        '''        
        regex = re.search("PARC\s\d\d/\d\d", obj['desc']);
        if regex != None:
            installmentNumber = int(regex.group()[5:7])
            originalDate = datetime.strptime(obj['date'], '%Y%m%d')
            newDate = originalDate + relativedelta(months=installmentNumber-1)

            obj['date'] = newDate.strftime('%Y%m%d')
            obj['desc'] = obj['desc'] + " DT ORIG: " + originalDate.strftime('%d/%m')

            print 'Updated installment transaction date. Installment Number: {0} Original: {1} Updated: {2}'.format(installmentNumber, originalDate, newDate)




bbvisa2ofx/bbvisa2ofx.py

from datetime import datetime
from txtparser import TxtParser

def convert ( fileTxt, fileOfx, closeOfxFile=True):
    """
        fileTxt: python file para o arquivo txt disponibilizado pelo banco do brasil
        fileOfx: python file com o caminho para o ofx que sera gerado
        closeOfxFile: se false nao fecha fileOfx (workaroud a classe para usar StringIO ao inves da classe file do python)

        Em um arquivo OFX temos um banco, que possui contas (neste caso apenas uma) que por sua vez possuem transacoes.
        Os itens retornados pelo parser do txt, representam transacoes de uma conta

        Para cada transacao o parametro FITID eh composto dos valores %(date)%(value)s%(desc) - este item eh preenchido para que o gnucash
        possa defirnir se o item jah foi importado ou nao
    """

    parser = TxtParser(fileTxt)
    parser.parse()
    items = parser.items
    cardTitle = parser.cardTitle
    cardNumber = parser.cardNumber

    today = datetime.now().strftime('%Y%m%d')


    # output
    out=fileOfx

    out.write (
        """OFXHEADER:100
DATA:OFXSGML
VERSION:102
SECURITY:NONE
ENCODING:USASCII
CHARSET:1252
COMPRESSION:NONE
OLDFILEUID:NONE
NEWFILEUID:NON

<OFX>
    <SIGNONMSGSRSV1>
        <SONRS>
            <STATUS>
                <CODE>0</CODE>
                <SEVERITY>INFO</SEVERITY>
            </STATUS>
            <DTSERVER>%(DTSERVER)s</DTSERVER>
            <LANGUAGE>POR</LANGUAGE>
            <FI>
                <ORG>Banco do Brasil</ORG>
                <FID>1</FID>
            </FI>
        </SONRS>
    </SIGNONMSGSRSV1>
    <BANKMSGSRSV1>
        <STMTTRNRS>
            <TRNUID>1</TRNUID>
            <STATUS>
                <CODE>0</CODE>
                <SEVERITY>INFO</SEVERITY>
            </STATUS>
            <STMTRS>
                <CURDEF>BRL</CURDEF>
                <BANKACCTFROM>
                   <BANKID>%(BANKID)s</BANKID>
                   <ACCTID>%(ACCTID)s</ACCTID>
                   <ACCTTYPE>CHECKING</ACCTTYPE>
                </BANKACCTFROM>
                <BANKTRANLIST>
                    <DTSTART>%(DTSERVER)s</DTSTART>
                    <DTEND>%(DTSERVER)s</DTEND>
        """ % {'DTSERVER':today,'BANKID':cardTitle.replace(' ',''),'ACCTID':cardNumber}
    )

    for item in items:
        out.write(
                """
                    <STMTTRN>
                        <TRNTYPE>OTHER</TRNTYPE>
                        <DTPOSTED>%(date)s</DTPOSTED>
                        <TRNAMT>%(value)s</TRNAMT>
                        <FITID>%(fitid)s</FITID>
                        <MEMO>%(desc)s</MEMO>
                    </STMTTRN>""" % item
                  )



    out.write(
        """
                </BANKTRANLIST>
                <LEDGERBAL>
                    <BALAMT>0</BALAMT>
                    <DTASOF>%s</DTASOF>
                </LEDGERBAL>
            </STMTRS>
        </STMTTRNRS>
    </BANKMSGSRSV1>
</OFX>
        """ % today


    )

    if(closeOfxFile):
        out.close()
    print "Convertido com sucesso!"




src/bbvisa2ofx/txtparsertest.py
'''
Created on Jun 9, 2010

@author: coelho
'''
import unittest
from txtparser import TxtParser
from bbvisa2ofx import convert

class Test(unittest.TestCase):
    file_path = "../testfiles/exemploFaturaCartao.txt";
    parser = ''

    def setUp(self):
        fTxt = open(self.file_path,'r')
        
        self.parser = TxtParser(fTxt);

    def testParse(self):
        self.parser.parse()
        print self.parser.items    

    def testParse_ConvertingDollarToReal(self):
        self.parser.parse()
        
        self.assertEquals(-978.26324599999998, self.parser.items[6]['value'])

    def testConvert(self):
        fTxt = open(self.file_path,'r')
        fOfx = open(self.file_path+".ofx",'w')
        
        convert(fTxt,fOfx)
    
    def testIsDolarLine(self):
        self.assertEquals(self.parser.parseExchangeRateLine('      0,00 -       0,00 +       0,00 =        0,00   X      0.0 =           0,00'),0.0)
        self.assertEquals(self.parser.parseExchangeRateLine('    871,64 -      21,78 +      69,52 =      823,90   X   1.7627 =       1.452,28'),1.7627)
        self.assertEquals(self.parser.parseExchangeRateLine('  1.172,33 -   1.172,33 +   1.039,98 =    1.039,98 -       0,00 =       1.039,98'),0.0)
        self.assertEquals(self.parser.parseExchangeRateLine('    214,17 -     214,79 +     162,00 =      161,38 -       0,00 =         161,3'),0.0)
        self.assertEquals(self.parser.parseExchangeRateLine('11/08/10 TARIFA SOBRE COMPRAS NO EXTERIOR                       0,00        0,44'),0.0)
        self.assertEquals(self.parser.parseExchangeRateLine(''),0.0)
        
if __name__ == "__main__":
    #import sys;sys.argv = ['', 'Test.testName']
    unittest.main()